### DISCUSSION TITLE
[week17] 作業一 Sequelize 資料庫關聯

### DISCUSSION THREAD

**Discussion by chengcheng1231** [Lidemy/forum#211] [其他問題] [answered]
困擾我很多天的問題，有請路過高人指點 ><，感謝
1. 我已經利用 sequelize-cli 來建立兩個資料表，chencheng_hw17_contents 有包含 UserId
![image](https://user-images.githubusercontent.com/36998811/134817332-29eb5fba-2e6f-4632-9c50-323f0ad21fec.png)
![image](https://user-images.githubusercontent.com/36998811/134817353-66642c73-60a8-4e51-ac0b-f8ab0978cca5.png)

2. 我也在 models 底下的檔案建立關聯
```javascript
// models/chengcheng_hw17.contents.js
'use strict';
const {
  Model
} = require('sequelize');
module.exports = (sequelize, DataTypes) => {
  class chengcheng_hw17_contents extends Model {
    static associate(models) {
      chengcheng_hw17_contents.belongsTo(models.chengcheng_hw17_users);
    }
  };
  chengcheng_hw17_contents.init({
    username: DataTypes.STRING,
    article_titile: DataTypes.STRING,
    article_content: DataTypes.TEXT,
    article_category: DataTypes.STRING,
    is_deleted: DataTypes.INTEGER,
  }, {
    sequelize,
    modelName: 'chengcheng_hw17_contents',
  });
  return chengcheng_hw17_contents;
};
```
```javascript
// models/chengcheng_hw17.users.js
'use strict';
const {
  Model
} = require('sequelize');
module.exports = (sequelize, DataTypes) => {
  class chengcheng_hw17_users extends Model {
    static associate(models) {
      // define association here
      chengcheng_hw17_users.hasMany(models.chengcheng_hw17_contents);
    }
  };
  
  chengcheng_hw17_users.init({
    username: DataTypes.STRING,
    password: DataTypes.STRING,
    about: DataTypes.STRING
  }, {
    sequelize,
    modelName: 'chengcheng_hw17_users',
  });
  return chengcheng_hw17_users;
};
```
3. 但當執行 index.js 並傳給 controller 來操作的時候，沒辦法用 findAll 成功去拿資料並印出來看看
```javascript
// controllers/content.js
const db = require('../models')
const Cotents = db.chengcheng_hw17_contents
const User = db.chengcheng_hw17_users

const contentController = {
    index: (req, res) => {
        Cotents.findAll({
            include: User,
        }).then(cotents => {
            console.log(JSON.stringify(cotents))
            res.render('front')
        })
    },
}

module.exports = contentController
```
4. 後來我發現程式幫我下的 SQL 不是用預期的 UserId 關聯而是 chengchengHw17UserId，所以我嘗試把資料庫結構名稱 UserId 改成 chengchengHw17UserId 就可以拿到資料，但是很奇怪，從頭到尾都沒有定義過要用 chengchengHw17UserId 來關聯，請問是哪裡可以改回來用 UserId 來關聯就好嗎

這是 terminal 印錯的資訊：
![image](https://user-images.githubusercontent.com/36998811/134817953-36eebf5e-dcbc-4cc9-a723-883604488ce9.png)

**Comment by allenliaooooooo** [comment, marked as answer]
定義關聯的時候如果沒有指定 foreign key 的 name，預設就會是Model name + id，也就是你的 Model name -> chengchengHw17User + Id = chengchengHw17UserId

老師有說過儘量不要改 sequelize 預設的欄位名稱，不然可能會遇到一些預期外的錯誤，真的要改的話好像可以在 define association 的地方指定 foreign key 再用 as 另外取名，詳細可以再看看官方文件

**Comment by chengcheng1231** [comment]
感謝 @rockyooooooo ，成功關聯起來了 🙇，原來它會預設成 Model name + id。
但自己測試的時候以為在 + id 這個步驟之前一定會將你的 Model name 做判斷並且加工，如果結尾是 s 就會幫你把 s 拿掉，"_" 也會拿掉改成 camelcase ，但事實上也不一定 !!!，現在還不太曉得它的轉換規則是什麼，可能真的要查文件了，所以最穩的命名還是將 User 當作 Model name ，讓它去 + id 變成 UserId 就好，和當初產生的欄位 UserId 一樣才關聯的起來，不然第二種方式就是要看 terminal 回傳它抓不到什麼關聯的 Model name + id，去改 migrations 裡面定義的欄位，再跑一次 db:migrate 更新資料表，讓兩個相符才關聯的起來
